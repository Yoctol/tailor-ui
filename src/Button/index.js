import PropTypes from 'prop-types';
import React, { PureComponent, createRef } from 'react';
import styled, { css, keyframes } from 'styled-components';
import { ifProp, switchProp } from 'styled-tools';
import { space, themeGet } from 'styled-system';

import controlTransition from '../utils/transition';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const StyledButton = styled.button`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${themeGet('borders.default')};
  border-radius: ${themeGet('radii.1')};
  border-color: ${themeGet('colors.primary')};
  background-color: ${themeGet('colors.primaryDark')};
  color: ${themeGet('colors.light')};
  line-height: ${themeGet('lineHeight')};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    border-color: ${themeGet('colors.primaryDark')};
    outline: 0;
  }

  &:hover {
    background-color: ${themeGet('colors.light')};
    color: ${themeGet('colors.primaryDark')};
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  ${switchProp('size', {
    sm: css`
      padding: ${themeGet('space.paddingYSm')} ${themeGet('space.paddingXSm')};
      font-size: ${themeGet('fontSizes.sm')};
    `,
    m: css`
      padding: ${themeGet('space.paddingY')} ${themeGet('space.paddingX')};
      font-size: ${themeGet('fontSizes.default')};
    `,
    lg: css`
      padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
      font-size: ${themeGet('fontSizes.lg')};
    `,
  })}

  ${ifProp(
    'block',
    css`
      width: 100%;
    `
  )}

  ${ifProp(
    'danger',
    css`
      border-color: ${themeGet('colors.error')};
      background-color: ${themeGet('colors.error')};
      color: ${themeGet('colors.light')};

      &:focus {
        border-color: ${themeGet('colors.error')};
      }

      &:hover {
        background-color: ${themeGet('colors.error')};
        color: ${themeGet('colors.light')};
      }
    `
  )};

  ${ifProp(
    'light',
    css`
      border-color: ${themeGet('colors.gray.8')};
      background-color: ${themeGet('colors.light')};
      color: ${themeGet('colors.gray.2')};

      &:focus {
        border-color: ${themeGet('colors.primaryDark')};
      }

      &:hover {
        background-color: ${themeGet('colors.gray.8')};
      }

      ${ifProp(
        'active',
        css`
          border-color: ${themeGet('colors.secondaryDark')};
          background-color: ${themeGet('colors.secondaryDark')};
          color: ${themeGet('colors.light')};

          &:hover {
            background-color: ${themeGet('colors.secondary')};
            color: ${themeGet('colors.light')};
          }

          &:focus {
            border-color: ${themeGet('colors.secondaryDark')};
          }
        `
      )};
    `
  )};

  ${ifProp(
    'circle',
    css`
    border-radius: 999px;

    ${ifProp(
      { size: 'sm' },
      css`
        padding: ${themeGet('space.paddingYSm')}
          calc(${themeGet('space.paddingXSm')} * 2);
      `
    )}
    ${ifProp(
      { size: 'm' },
      css`
        padding: ${themeGet('space.paddingY')}
          calc(${themeGet('space.paddingX')} * 2);
      `
    )}
    ${ifProp(
      { size: 'lg' },
      css`
        padding: ${themeGet('space.paddingYLg')}
          calc(${themeGet('space.paddingXLg')} * 2);
      `
    )}
  `
  )}

  ${ifProp(
    'ghost',
    css`
      border-color: ${themeGet('colors.light')};
      background-color: transparent;
    `
  )}

  ${ifProp(
    'loading',
    css`
      color: transparent;
      pointer-events: none;

      &::after {
        content: '';
        display: block;
        position: absolute;
        top: calc(50% - (1em / 2));
        left: calc(50% - (1em / 2));
        width: 1em;
        height: 1em;
        border: 2px solid ${themeGet('colors.gray.6')};
        border-radius: 50%;
        border-top-color: transparent;
        border-right-color: transparent;
        animation: ${spin} 0.7s linear infinite;
      }
    `
  )}

  ${controlTransition()};

  ${space};
`;

class Button extends PureComponent {
  constructor(props) {
    super(props);
    this.ripple = createRef();
    this.button = createRef();
  }

  handleClick = e => this.ripple.current.startRipple(e, this.button);

  render() {
    const { children, ...props } = this.props;
    return (
      <StyledButton
        innerRef={this.button}
        onMouseUp={this.handleClick}
        onTouchend={this.handleClick}
        {...props}
      >
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

Button.displayName = 'Button';

Button.propTypes = {
  /**
   * Description of prop "active".
   */
  active: PropTypes.bool,
  /**
   * Description of prop "block".
   */
  block: PropTypes.bool,
  /**
   * Description of prop "circle".
   */
  circle: PropTypes.bool,
  /**
   * Description of prop "ghost".
   */
  danger: PropTypes.bool,
  /**
   * Description of prop "light".
   */
  ghost: PropTypes.bool,
  /**
   * Description of prop "danger".
   */
  light: PropTypes.bool,
  /**
   * Description of prop "loading".
   */
  loading: PropTypes.bool,
  /**
   * Description of prop "size".
   */
  size: PropTypes.oneOf(['sm', 'm', 'lg']),
  /**
   * Description of prop "Space".
   */
  ...space.propTypes,
};

Button.defaultProps = {
  size: 'm',
  block: false,
  circle: false,
  danger: false,
  ghost: false,
  light: false,
  active: false,
  loading: false,
};

export default Button;
