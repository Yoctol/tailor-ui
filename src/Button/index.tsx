import React, {
  MouseEvent,
  PureComponent,
  ReactInstance,
  RefObject,
  createRef,
} from 'react';
import { SpaceProps, space as styledSpace } from 'styled-system';

import styled, { css, keyframes } from 'utils/styled-components';
import tag from 'utils/CleanTag';

import Icon, { IconType, IconWrapper } from '../Icon';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

type ButtonType =
  | 'default'
  | 'primary'
  | 'secondary'
  | 'info'
  | 'success'
  | 'warning'
  | 'danger';

export interface IGetTypesStylesInterface {
  type?: ButtonType;
  text?: boolean;
  outlined?: boolean;
}

const getTypeStyles = ({
  type = 'default',
  text,
  outlined,
  theme,
}: IGetTypesStylesInterface & {
  theme: any;
}) => {
  const { colors } = theme;

  if (text) {
    return {
      borderColor: 'transparent',
      backgroundColor: 'transparent',
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray300,
        borderColor: 'transparent',
      },
    };
  }

  if (type === 'default') {
    return {
      borderColor: colors.gray300,
      backgroundColor: colors.light,
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray200,
        borderColor: colors.primary,
      },
    };
  }

  const bg = colors[type];
  const { light } = colors;
  const backgroundColor = outlined ? 'transparent' : bg;

  return {
    borderColor: bg,
    backgroundColor,
    color: outlined ? bg : light,
    hover: {
      backgroundColor: outlined ? bg : light,
      color: outlined ? light : bg,
    },
  };
};

interface IGetLoadingInterface extends IGetTypesStylesInterface {
  loading?: boolean;
}

const getLoading = ({
  loading,
  ...props
}: IGetLoadingInterface & {
  theme: any;
}) =>
  loading &&
  css`
    opacity: 0.7;
    cursor: default;
    pointer-events: none;

    &::before {
      width: 1em;
      height: 1em;
      margin-right: 8px;
      border: 1px solid ${getTypeStyles(props).color};
      border-radius: 50%;
      border-top-color: transparent;
      border-right-color: transparent;
      animation: ${spin} 0.7s linear infinite;
    }
  `;

type ButtonSize = 'sm' | 'md' | 'lg';

interface IGetSizeInterface {
  size: ButtonSize;
  hasIcon?: boolean;
}

const getSize = css`
  ${({
    hasIcon,
    size,
    theme: { paddings, heights, fontSizes },
  }: IGetSizeInterface & { theme: any }) => {
    if (hasIcon) {
      return {
        sm: css`
          width: ${heights.sm};
          height: ${heights.sm};
        `,
        md: css`
          width: ${heights.base};
          height: ${heights.base};
        `,
        lg: css`
          width: ${heights.lg};
          height: ${heights.lg};
        `,
      }[size];
    }

    return {
      sm: css`
        height: ${heights.sm};
        padding: 0 ${paddings.sm};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 0 ${paddings.md};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 0 ${paddings.lg};
        font-size: ${fontSizes.lg};
      `,
    }[size];
  }};
`;

interface IBlockInterface {
  block?: boolean;
}

const getBlock = ({ block }: IBlockInterface) =>
  block &&
  css`
    width: 100%;
  `;

interface IRoundedInterface {
  rounded?: boolean;
}

const getRounded = ({ rounded }: IRoundedInterface) =>
  rounded &&
  css`
    border-radius: 999px;
  `;

const getTypes = ({
  type = 'default',
  text = false,
  outlined = false,
  theme,
}: IGetTypesStylesInterface & { theme: any }) => {
  const styles = getTypeStyles({ type, text, outlined, theme });

  return css`
    border-color: ${styles.borderColor};
    background-color: ${styles.backgroundColor};
    color: ${styles.color};

    ${IconWrapper /* sc-selector */} {
      cursor: pointer;
      svg {
        fill: ${styles.color};
      }
    }

    &:hover {
      border-color: ${styles.hover.borderColor};
      background-color: ${styles.hover.backgroundColor};
      color: ${styles.hover.color};

      ${IconWrapper /* sc-selector */} svg {
        fill: ${styles.hover.color};
      }
    }
  `;
};

type StyledButtonProps = IGetTypesStylesInterface &
  IGetLoadingInterface &
  IGetSizeInterface &
  IBlockInterface &
  IRoundedInterface &
  SpaceProps;

export const StyledButton = styled<StyledButtonProps, 'button'>(tag.button)`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  line-height: ${p => p.theme.lineHeight};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    outline: 0;
  }

  &:disabled,
  &[disabled] {
    opacity: 0.7;
    cursor: default;
    pointer-events: none;
  }

   &::before {
      content: '';
      display: inline-block;
      position: relative;
      left: 0;
      width: 0;
      height: 0;
      margin-right: 0;
      ${p => p.theme.transition /* sc-declaration */};
    }

  ${getBlock /* sc-declaration */}
  ${getRounded /* sc-declaration */}
  ${getSize /* sc-declaration */}
  ${getTypes /* sc-declaration */}
  ${getLoading /* sc-declaration */};

  ${p => p.theme.transition /* sc-declaration */};
  ${styledSpace};
`;

export type ButtonProps = SpaceProps & {
  type?: ButtonType;
  text?: boolean;
  block?: boolean;
  rounded?: boolean;
  outlined?: boolean;
  loading?: boolean;
  icon?: string | IconType;
  size?: ButtonSize;
  onClick?: (event: MouseEvent) => void;
};

class Button extends PureComponent<ButtonProps> {
  ripple: RefObject<Ripple> = createRef<Ripple>();

  buttonRef: ReactInstance | null = null;

  handleButtonRef = (ref: any) => {
    this.buttonRef = ref;
  };

  handleClick = (event: MouseEvent) => {
    if (this.ripple.current && this.buttonRef) {
      this.ripple.current.startRipple(event, this.buttonRef);
    }
  };

  render() {
    const {
      children,
      icon,
      loading = false,
      type = 'default',
      size = 'md',
      ...props
    } = this.props;

    return (
      <StyledButton
        ref={this.handleButtonRef}
        onMouseUp={this.handleClick}
        hasIcon={Boolean(icon && !children)}
        loading={loading}
        size={size}
        type={type}
        {...props}
      >
        {!loading &&
          icon && (
            <Icon
              type={icon}
              size="20"
              mr={children ? 1 : 0}
              style={{ pointerEvents: 'none' }}
            />
          )}
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

export default Button;
