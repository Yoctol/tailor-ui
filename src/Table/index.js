import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  themeGet,
  width,
  space,
  textAlign,
  borders,
  borderColor,
} from 'styled-system';

import theme from '../theme';

const HeadColumn = styled.th`
  padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
  font-weight: 500;

  ${width};
  ${space};
  ${borders};
  ${borderColor};
`;

HeadColumn.propTypes = {
  ...width.propTypes,
  ...space.propTypes,
  ...borders.propTypes,
  ...borderColor.propTypes,
};

HeadColumn.defaultProps = {
  theme,
  borderColor: 'border',
};

const Column = styled.td`
  padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};

  ${borders};
  ${borderColor};
`;

Column.propTypes = {
  ...space.propTypes,
  ...borders.propTypes,
  ...borderColor.propTypes,
};

Column.defaultProps = {
  theme,
  borderColor: 'border',
};

const Row = styled.tr`
  border-bottom: ${themeGet('borders.default')} ${themeGet('colors.border')};
`;

Row.defaultProps = {
  theme,
};

const Head = ({ children }) => (
  <thead>
    <Row>{children}</Row>
  </thead>
);

Head.propTypes = {
  children: PropTypes.element.isRequired,
};

const Body = ({ children }) => <tbody>{children}</tbody>;

Body.propTypes = {
  children: PropTypes.element,
};

Body.defaultProps = {
  children: '',
};

const Table = styled.table`
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: ${themeGet('radii.1')};
  border-top-right-radius: ${themeGet('radii.1')};
  box-shadow: 0 0 0 1px ${themeGet('colors.border')};

  & > thead {
    border-bottom: ${themeGet('borders.default')} ${themeGet('colors.border')};
    background-color: ${themeGet('colors.gray.9')};
  }

  ${textAlign};
  ${width};
`;

Table.defaultProps = {
  theme,
  width: '100%',
  textAlign: 'left',
};

Table.Head = Head;
Table.HeadColumn = HeadColumn;
Table.Body = Body;
Table.Row = Row;
Table.Column = Column;

export default Table;
