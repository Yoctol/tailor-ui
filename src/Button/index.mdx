---
name: Button
menu: General
route: /general/button
---

import { PropsTable, Playground } from 'docz';
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
import { Toggle } from 'react-powerplug';

import Box from '../Grid/Box';
import Icon from '../Icon';

import Button from './';

# Button

To trigger an operation.

## When To Use

A button means an operation (or a series of operations). Clicking a button will trigger corresponding business logic.

## Examples

```js
import { Button } from '@yoctol/ui';
```

### Basic

<Playground>
  <Button>Basic Button</Button>
  <Button text ml="2">Text Button</Button>
</Playground>

### Type

<Playground>
  <Button type="primary">Primary</Button>
  <Button type="secondary" ml="2">
    Secondary
  </Button>
  <Button type="info" ml="2">
    Info
  </Button>
  <Button type="success" ml="2">
    Success
  </Button>
  <Button type="warning" ml="2">
    Warning
  </Button>
  <Button type="danger" ml="2">
    Danger
  </Button>
</Playground>

### Rounded

<Playground>
  <Button type="primary" rounded>Primary</Button>
  <Button type="secondary" ml="2" rounded>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded>
    Info
  </Button>
  <Button type="success" ml="2" rounded>
    Success
  </Button>
  <Button type="warning" ml="2" rounded>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded>
    Danger
  </Button>
</Playground>

### Outlined

<Playground>
  <Button type="primary" outlined>Primary</Button>
  <Button type="secondary" ml="2" outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" outlined>
    Info
  </Button>
  <Button type="success" ml="2" outlined>
    Success
  </Button>
  <Button type="warning" ml="2" outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" outlined>
    Danger
  </Button>
</Playground>

### Rounded & Outlined

<Playground>
  <Button type="primary" rounded outlined>Primary</Button>
  <Button type="secondary" ml="2" rounded outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded outlined>
    Info
  </Button>
  <Button type="success" ml="2" rounded outlined>
    Success
  </Button>
  <Button type="warning" ml="2" rounded outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded outlined>
    Danger
  </Button>
</Playground>

### Icon

```js
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
```

<Playground>
  <Button type="primary" icon={MdCheck}>Confirm</Button>
  <Button type="danger" icon={MdDelete} ml="2">Delete</Button>
  <Button type="info" icon={MdSearch} ml="2">Search</Button>
  <br />
  <br />
  <Button type="primary" rounded icon={MdSearch} ml="2" />
  <Button type="danger" rounded icon={MdDelete} ml="2" />
  <Button type="info" rounded icon={MdSearch} ml="2" />
  <br />
  <br />
  <Button>
    <Icon type={MdCheck} fill="success" size="20" mr="2" />
    Success
  </Button>
  <Button ml="2">
    <Icon type={MdClose} fill="danger" size="20" mr="2" />
    Danger
  </Button>
</Playground>


### Disabled

<Playground>
  <Button type="primary" disabled>Primary</Button>
  <Button type="secondary" ml="2" disabled>
    Secondary
  </Button>
  <Button type="info" ml="2" disabled>
    Info
  </Button>
  <Button type="success" ml="2" disabled>
    Success
  </Button>
  <Button type="warning" ml="2" disabled>
    Warning
  </Button>
  <Button type="danger" ml="2" disabled>
    Danger
  </Button>
</Playground>

### Loading

<Playground>
  <Button type="primary" loading>Primary</Button>
  <Button type="secondary" ml="2" loading>
    Secondary
  </Button>
  <Button type="info" ml="2" loading>
    Info
  </Button>
  <Button type="success" ml="2" loading>
    Success
  </Button>
  <Button type="warning" ml="2" loading>
    Warning
  </Button>
  <Button type="danger" ml="2" loading>
    Danger
  </Button>
  <br />
  <br />
  <Toggle>
    {({ on, toggle }) => (
      <Button
        loading={on}
        onClick={() => {
          toggle()
          setTimeout(toggle, 3000)
        }}
      >
        Click me!
      </Button>
    )}
  </Toggle>
</Playground>

### Sizes

<Playground>
  <Button size="sm">
    Button sm
  </Button>

  <Button ml={2}>
    Button md
  </Button>

  <Button ml={2} size="lg">
    Button lg
  </Button>
</Playground>

### Block

<Playground>
  <Button block>
    Block Button
  </Button>
</Playground>

## API

<PropsTable of={Button} />
