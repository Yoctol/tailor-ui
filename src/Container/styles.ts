import styled from 'styled-components';

import { Flex } from '../Layout';
import { StyledTableWrapper } from '../Table/styles';

export const StyledContainer = styled(Flex)`
  flex-direction: column;
  overflow: hidden;
  border-radius: ${(p) => p.theme.radii.xl};
  background-color: ${(p) => p.theme.colors.light};
  box-shadow: ${(p) => p.theme.shadows.base};
  color: ${(p) => p.theme.colors.gray700};
  font-size: ${(p) => p.theme.fontSizes.base};

  ${StyledTableWrapper} {
    width: 100%;
    border-top: ${(p) => p.theme.borders.base};
    border-bottom: ${(p) => p.theme.borders.base};
    border-radius: 0;
    border-color: ${(p) => p.theme.colors.gray300};
    box-shadow: none;
  }

  & + & {
    margin-top: 16px;
  }
`;

export const StyledSection = styled(Flex)`
  flex-direction: column;

  & + & {
    margin-top: 32px;
  }
`;
