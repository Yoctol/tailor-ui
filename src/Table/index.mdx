---
name: Table
menu: Components
route: /components/table
---

import { PropsTable, Playground } from 'docz';
import { range } from 'ramda';
import Table from './';

# Table

A table displays rows of data.

## When To Use
- To display a collection of structured data.
- To sort, search, paginate, filter data.

## Examples

```js
import { Table } from '@yoctol/ui';
```

### Basic Table

<Playground>
  <Table>
    <Table.Head>
      {range(2, 10).map(x => <Table.HeadColumn key={x}>{x}</Table.HeadColumn>)}
    </Table.Head>
    <Table.Body>
      {range(1, 10).map(x => (
        <Table.Row key={x}>
          {range(2, 10).map(y => (
            <Table.Column key={y}>
              {y} * {x} = {x * y}
            </Table.Column>
          ))}
        </Table.Row>
      ))}
    </Table.Body>
  </Table>
</Playground>

### Custom Table

<Playground>
  <Table textAlign="center">
    <Table.Head>
      <Table.HeadColumn colSpan="8">Time</Table.HeadColumn>
    </Table.Head>
    <Table.Body>
      <Table.Row>
        <Table.Column />
        <Table.Column>日</Table.Column>
        <Table.Column>一</Table.Column>
        <Table.Column>二</Table.Column>
        <Table.Column>三</Table.Column>
        <Table.Column>四</Table.Column>
        <Table.Column>五</Table.Column>
        <Table.Column>六</Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Morning<br />10:00-13:00
        </Table.Column>
        <Table.Column
          rowSpan="3"
          borderColor="gray300"
          borderLeft="base"
          borderRight="base"
        >
          Day off
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Afternoon<br />14:00-18:00
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Night<br />19:00-22:00
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
    </Table.Body>
  </Table>
</Playground>

## API

<PropsTable of={Table} />
