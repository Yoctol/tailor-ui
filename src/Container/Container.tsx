import React, {
  ComponentPropsWithoutRef,
  FC,
  ForwardRefExoticComponent,
  PropsWithoutRef,
  ReactNode,
  RefAttributes,
  forwardRef,
  isValidElement,
} from 'react';

import { Box, BoxProps } from '../Layout/Box';
import { Divider } from '../Divider';
import { Heading, Text } from '../Typography';

import { StyledContainer, StyledSection } from './styles';

export interface SectionProps extends Omit<BoxProps, 'color'> {
  title?: ReactNode;
  children: ReactNode;
}

const Section = forwardRef<HTMLDivElement, SectionProps>(function Section(
  { title, children, ...props },
  ref
) {
  return (
    <StyledSection ref={ref} {...props}>
      {title && <Heading.H5 mb="8px">{title}</Heading.H5>}
      {children}
    </StyledSection>
  );
});

const isElementTable = (element: ReactNode) =>
  isValidElement(element) &&
  ((element.type as unknown) as { displayName: string }).displayName ===
    'Table';

interface ContainerTitleProps {
  title?: ReactNode;
  subTitle?: ReactNode;
}

const ContainerTitle: FC<ContainerTitleProps> = ({ title, subTitle }) => {
  return (
    <Box pt="24px" px="32px">
      {title && <Heading.H4 mb="8px">{title}</Heading.H4>}
      {subTitle && (
        <Box mb="8px">
          <Text color="gray500" fontSize="sm">
            {subTitle}
          </Text>
        </Box>
      )}

      <Divider mb="24px" />
    </Box>
  );
};

interface ContainerChildrenProps {
  hasTitle: boolean;
}

const ContainerChildren: FC<ContainerChildrenProps> = ({
  hasTitle,
  children,
}) => {
  const hasChildrenTable = Array.isArray(children)
    ? children.some(isElementTable)
    : isElementTable(children);

  if (!hasChildrenTable) {
    return (
      <Box px="32px" pb="32px" pt={hasTitle ? '0' : '24px'}>
        {children}
      </Box>
    );
  }

  if (!Array.isArray(children)) {
    return <>{children}</>;
  }

  const childrenIndex =
    (Array.isArray(children) &&
      children.findIndex((child) => isElementTable(child))) ||
    -1;

  return (
    <>
      {children.slice(0, childrenIndex).length > 0 && (
        <Box py="24px" px="32px">
          {children.slice(0, childrenIndex)}
        </Box>
      )}

      {children[childrenIndex]}

      {children.slice(childrenIndex + 1).length > 0 && (
        <Box py="24px" px="32px">
          {children.slice(childrenIndex + 1)}
        </Box>
      )}
    </>
  );
};

export type ContainerProps = Omit<BoxProps, 'color'> &
  ContainerTitleProps &
  ComponentPropsWithoutRef<'div'>;

const Container = forwardRef<HTMLDivElement, ContainerProps>(function Container(
  { title, subTitle, children, ...props },
  ref
) {
  const hasTitle = Boolean(title || subTitle);

  return (
    <StyledContainer ref={ref} {...props}>
      {hasTitle && <ContainerTitle title={title} subTitle={subTitle} />}
      <ContainerChildren hasTitle={hasTitle}>{children}</ContainerChildren>
    </StyledContainer>
  );
}) as ForwardRefExoticComponent<
  PropsWithoutRef<ContainerProps> & RefAttributes<HTMLDivElement>
> & {
  Section: typeof Section;
};

Container.Section = Section;

export { Container };
