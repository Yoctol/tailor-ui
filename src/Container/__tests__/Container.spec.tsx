import React from 'react';

import { render } from 'test/test-utils';

import { Button } from '../../Button';
import { Container } from '../Container';
import { Heading } from '../../Typography';
import { Table } from '../../Table';

describe('Container', () => {
  it('should render correctly', () => {
    const { container } = render(
      <div>
        <Container title="Title A">
          <Container.Section title="Subtitle #1">
            Subtitles are text derived from either a transcript or screenplay of
            the dialog or commentary in films.
          </Container.Section>
          <Container.Section title="Subtitle #2">
            Subtitles are text derived from either a transcript or screenplay of
            the dialog or commentary in films.
          </Container.Section>
        </Container>

        <Container title={<>Title B</>} subTitle={<>Sub Title B</>}>
          <Container.Section title="Subtitle #1">
            Subtitles are text derived from either a transcript or screenplay of
            the dialog or commentary in films.
          </Container.Section>
          <Container.Section title="Subtitle #2">
            Subtitles are text derived from either a transcript or screenplay of
            the dialog or commentary in films.
          </Container.Section>
        </Container>
      </div>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render without title correctly', () => {
    const { container } = render(
      <Container>The content of container</Container>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with table correctly', () => {
    const { container } = render(
      <Container>
        <Heading.H4>Text Here</Heading.H4>

        <Table>
          <Table.Head>
            <Table.HeadColumn>Name</Table.HeadColumn>
            <Table.HeadColumn>Platform</Table.HeadColumn>
            <Table.HeadColumn>People</Table.HeadColumn>
            <Table.HeadColumn>Ammount</Table.HeadColumn>
            <Table.HeadColumn>Action</Table.HeadColumn>
          </Table.Head>
          <Table.Body>
            <Table.Row>
              <Table.Column>Name</Table.Column>
              <Table.Column>Platform</Table.Column>
              <Table.Column>10,000</Table.Column>
              <Table.Column>NTD 1,000</Table.Column>
              <Table.Column>
                <Button variant="primary-invert">Action</Button>
              </Table.Column>
            </Table.Row>
          </Table.Body>
        </Table>

        <Button variant="regular">+ Button</Button>
      </Container>
    );

    expect(container).toMatchSnapshot();
  });
});
