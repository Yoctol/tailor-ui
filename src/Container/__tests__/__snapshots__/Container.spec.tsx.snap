// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container should render correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
}

.c3 {
  font-size: 1.125rem;
  margin: 0;
  color: #283040;
  font-weight: bold;
  line-height: 1.5;
  margin-bottom: 8px;
}

.c7 {
  font-size: 1rem;
  margin: 0;
  color: #283040;
  font-weight: bold;
  line-height: 1.5;
}

.c4 {
  background-color: #c4cfdd;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 1rem 0;
  clear: both;
  margin-top: 0.5rem;
  margin-bottom: 24px;
  margin-left: 0rem;
  margin-right: 0rem;
  color: #97a4ba;
  font-size: 0.875rem;
  font-weight: 500;
}

.c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 24px 32px 32px;
  border-radius: 1rem;
  background-color: #ffffff;
  box-shadow: 0 2px 4px 0 #e0e0e0;
  color: #283040;
  font-size: 1rem;
}

.c0 + .c0 {
  margin-top: 16px;
}

.c6 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 + .c5 {
  margin-top: 32px;
}

.c8 {
  margin-bottom: 12px;
}

<div>
  <div>
    <div
      class="c0 c1 c2"
    >
      <h4
        class="c3"
      >
        Title A
      </h4>
      <div
        class="c4"
        color="gray400"
        font-size="sm"
        font-weight="500"
        type="horizontal"
      />
      <div
        class="c5 c1 c6"
      >
        <h5
          class="c7 c8"
        >
          Subtitle #1
        </h5>
        Subtitles are text derived from either a transcript or screenplay of the dialog or commentary in films.
      </div>
      <div
        class="c5 c1 c6"
      >
        <h5
          class="c7 c8"
        >
          Subtitle #2
        </h5>
        Subtitles are text derived from either a transcript or screenplay of the dialog or commentary in films.
      </div>
    </div>
    <div
      class="c0 c1 c2"
    >
      <h4
        class="c3"
      >
        Title B
      </h4>
      <div
        class="c4"
        color="gray400"
        font-size="sm"
        font-weight="500"
        type="horizontal"
      />
      <div
        class="c5 c1 c6"
      >
        <h5
          class="c7 c8"
        >
          Subtitle #1
        </h5>
        Subtitles are text derived from either a transcript or screenplay of the dialog or commentary in films.
      </div>
      <div
        class="c5 c1 c6"
      >
        <h5
          class="c7 c8"
        >
          Subtitle #2
        </h5>
        Subtitles are text derived from either a transcript or screenplay of the dialog or commentary in films.
      </div>
    </div>
  </div>
</div>
`;
